optimisation/design/Third_Party_API_Design.md

```diff
@@ -0,0 +1,595 @@
+# An Application Programming Interface for user defined model functions in PACE
+**2020-05-15**
+
+
+# Scope
+
+This documents details the required function signatures for user defined sample scattering models
+and how they interface with Horace, using either Matlab, Python, C, C++ or Fortran.
+
+
+# Motivation
+
+Analysis of inelastic neutron scattering data almost always requires a parameterised model of the measured scattering intensities.
+Horace, as currently implemented, requires this model to be in the form of a Matlab function.
+This function is required to either 
+
+1. Take a 4D vector of momentum-energy ![\mathbb{Q} \equiv (\mathb{Q},E)] and yield the scattering function ![S(\mathbb{Q})].
+2. Take the 3D vector of momentum transfer ![\mathbf{Q}] only and yield ![N] sets of vectors ![(E_n, S(\mathbf{Q}, E_n))] where ![N] is the number of dispersion branches.
+
+For type 2, the ![N] sets of vectors are contained in a cell array
+and a helper function `disp2sqw` is used to obtain ![S(\mathbb{Q})]
+by convolving the calculated delta-function dispersion with broadening functions (Gaussian is default). 
+Whilst the internal fitting routines of Horace (including for resolution convolution) requires type 1 model functions (yielding ![S(\mathbb{Q})]),
+many actual user models are type 2, including `SpinW` and `Euphonic`.
+Thus, the new PACE implementation should allow both types of models.
+We also note, that when instrumental resolution convolution (`TobyFit`) is used,
+the broadening function in `disp2sqw` should only represent some intrinsic physical broadening
+(such as a quasiparticle lifetime).
+Whereas when `TobyFit` is not used,
+the broadening function in `disp2sqw` will be a combination of both intrinsic and instrumental resolution broadening.
+
+In the current Horace implementation, these model functions are used by the `multifit_sqw` function/class
+(which inherits from `multifit`) to fit to data using a built-in Levenberg-Marquardt optimizer. 
+The [new design for PACE](Model_Optimisation_Design.md) proposes instead to use a pair of classes, `OptFunction` and `OptModel`.
```

Comment:
```suggestion
The [new design for PACE](Model_Optimisation_Design.md) proposes instead to use a trio of classes, `OptFunction`, `OptFunctions`, and `OptModel`.
```
or since `OptFunctions` probably isn't relevant for this discussion 
```suggestion
The [new design for PACE](Model_Optimisation_Design.md) proposes instead to use a host of classes including `OptFunction` and `OptModel`.
```